### PR TITLE
Template Injection Emulator

### PR DESCRIPTION
**Pull request by rjt-gupta**
Engines Supported:

- Mako
- Tornado

All of them are detected by using separate regex and have separate `get_injection` methods.

### CODE REVIEW

**Comment by rjt-gupta** [comment]
We need to fix the travis failing build first.

File: tanner/emulators/template_injection.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    def scan(self, value):
+        detection = None
+
+        if patterns.TEMPLATE_INJECTION_JINJA2.match(value) or patterns.TEMPLATE_INJECTION_MAKO.match(value) or \
+                patterns.TEMPLATE_INJECTION_TORNADO.match(value):
+            detection = dict(name='template_injection', order=3)
+        return detection
+
+    async def handle(self, attack_params):
```
Comment (rnehra01) [inline tanner/emulators/template_injection.py]: It would be great if you could test the code and then push it at least once. It took me some time to figure out the error. `handle` usually needs 3 params in the base emulator, not 2.

File: tanner/emulators/template_injection.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    def __init__(self, loop=None):
+        self._loop = loop if loop is not None else asyncio.get_event_loop()
+        self.logger = logging.getLogger('tanner.template_injection')
+
+    def get_injection_result(self, payload):
+
+        template_injection_result = None
+        base_template = """<html><head><title>Welcome %s !!</title></head></html>""" % payload
+
+        if patterns.TEMPLATE_INJECTION_MAKO.match(payload):
```
Comment (rnehra01) [inline tanner/emulators/template_injection.py]: This one is giving me 500 using `?a=<%= 7*7 %>`, look into it. I guess there is some URL decoding issue.

**Comment by rnehra01** [comment]
Overall it looks pretty limited to me. I could get to work simple payloads like `{{7*7}}` which doesn't have serious impacts. Have you thought about complex payloads that can have high severity like reading local files, configs etc. https://github.com/swisskyrepo/PayloadsAllTheThings/tree/master/Server%20Side%20Template%20Injection#jinja2

**Comment by rjt-gupta** [comment]
> 
> 
> Overall it looks pretty limited to me. I could get to work simple payloads like `{{7*7}}` which doesn't have serious impacts. Have you thought about complex payloads that can have high severity like reading local files, configs etc. https://github.com/swisskyrepo/PayloadsAllTheThings/tree/master/Server%20Side%20Template%20Injection#jinja2

Yes, this emulator is working with payloads running commands and opening password files as well. I will push all the tests by today.

File: tanner/emulators/template_injection.py (hunk 1 above)
Comment (rjt-gupta) [reply tanner/emulators/template_injection.py]: Oh! I missed the `session` parameter. 😅

**Comment by rjt-gupta** [comment]
So, payloads running commands or reading `/etc/passwd` can be done easily for `mako`, `tornado` template engines. For jinja2 I thought a simple detection would work as its often detected using `{{7*7}}` methods.

Do you have something in mind for jinja2 engine?

File: tanner/emulators/template_injection.py (hunk 2 above)
Comment (rjt-gupta) [reply tanner/emulators/template_injection.py]: For injections in `mako` you have to use something like this- `<% x=7*7 %>${x}`.

File: tanner/tests/test_template_injection.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        self.assertIn(self.expected_result, self.returned_result)
+
+    def test_handle_jinja2(self):
+        payload = '{{ 7*"7" }}'
+
+        self.returned_result = self.handler.get_injection_result(payload)
+        self.expected_result = '7777777'
+        self.assertIn(self.expected_result, self.returned_result)
+
+    def test_handle_tornado(self):
```
Comment (rnehra01) [inline tanner/tests/test_template_injection.py]: So this can compromise can the entire honeypot and attack will soon understand that it is talking to a honeypot. Just think about why we're using Docker containers / phpox for `code execution` emulators.

File: tanner/emulators/template_injection.py (hunk 2 above)
Comment (rnehra01) [reply tanner/emulators/template_injection.py]: I would like to know how you test the emulators. It looks like you don't test payloads using snare but directly execute the `get_injection_result` method.

File: tanner/tests/test_template_injection.py (hunk 3 above)
Comment (rjt-gupta) [reply tanner/tests/test_template_injection.py]: I am just demonstrating the possibilities. In real production, all of this can be access controlled and also suitable filters can be applied to prevent such payloads.

Comment (rnehra01) [reply tanner/tests/test_template_injection.py]: I think you don't understand honeypot, its aim is to collect all sort of working payloads (simple+complex).

Comment (rnehra01) [reply tanner/tests/test_template_injection.py]: The emulator should emulate almost all the payloads.

**Comment by rjt-gupta** [comment]
So, the twig engine is working fine along with from snare side as well.
To run the emulator install twig using composer - `composer require "twig/twig:^2.0"` from '/tanner` directory.

It creates a folder `vendor/` with all the twig files. `autoload.php` etc

Input format - `?a={{7*"7"}}` 
Output should contain `49` which is twig engine specific.

**Comment by rjt-gupta** [comment]
Also, maybe we can include this `vendor/` files as a part of project then user won't have to worry about anything. What say?

**Comment by rnehra01** [comment]
@rjt-gupta What payloads are working for this emulator?

**Comment by rjt-gupta** [comment]
Engines - Mako, Tornado are working using docker and a custom image(base image - `alphine`).

Working complex payloads - 

Mako -  `<%\nimport os\nx=os.uname()\n%>\n${x}`
Tornado - `{%import os%}{{os.uname()}}`

(use `docker-compose build` before using this)

File: tanner/utils/patterns.py [hunk 4]
```diff
@@ -4,7 +4,6 @@
 RFI_ATTACK = re.compile(r'.*((http(s){0,1}|ftp(s){0,1}):).*', re.IGNORECASE)
 LFI_ATTACK = re.compile(r'.*(/\.\.)*(home|proc|usr|etc)/.*')
 LFI_FILEPATH = re.compile(r'((\.\.|/).*)')
-XSS_ATTACK = re.compile(r'.*<(.|\n)*?>')
```
Comment (rnehra01) [inline tanner/utils/patterns.py]: Why did you remove it from here?

File: tanner/emulators/template_injection.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                           ?>""" % (path_to_vendor, payload)
+
+        template_injection_result = await self.helper.get_result(twig_template)
+        return template_injection_result
+
+    def get_injection_result_docker(self, payload):
+        execute_result = None
+
+        client = docker.from_env()
+        docker_image = 'mushorg/template_injection:latest'
```
Comment (rnehra01) [inline tanner/emulators/template_injection.py]: I think you should use `docker_helper`.

File: docker/docker-compose.yml [hunk 6]
```diff
@@ -90,3 +90,15 @@ services:
      - tanner_api
      - tanner_web
      - tanner_phpox
+
+# Template Injection Emulator
```
Comment (rnehra01) [inline docker/docker-compose.yml]: Here docker compose is used to setup tanner as a whole. I think you should update `docker_helper.py` and add functionality related to [build](https://docker-py.readthedocs.io/en/stable/images.html#docker.models.images.ImageCollection.build), currently, we have only pull related [method](https://github.com/mushorg/tanner/blob/master/tanner/utils/docker_helper.py#L18).

File: docker/tanner/template injection/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+RUN apk -U --no-cache add \
+  python3
+
+RUN pip3 install --upgrade pip && \
+    pip3 install --no-cache-dir \
+    tornado \
+    jinja2 \
+    mako
+
+CMD ["python3", "-c", "import tornado"]
```
Comment (rnehra01) [inline docker/tanner/template injection/Dockerfile]: Is it required?

File: tanner/utils/docker_helper.py [hunk 8]
```diff
@@ -59,8 +59,8 @@ def __init__(self):
         execute_result = None
         try:
             container.start()
-            execute_result = container.exec_run(['sh', '-c', cmd]).decode('utf-8')
+            execute_result = container.exec_run(['sh', '-c', cmd])
             container.kill()
         except docker.errors.APIError as server_error:
             self.logger.exception('Error while executing command %s in container %s', cmd, server_error)
-        return execute_result
+        return execute_result.output.decode('utf-8')
```
Comment (rnehra01) [inline tanner/utils/docker_helper.py]: This is giving me error.

Comment (rnehra01) [reply tanner/utils/docker_helper.py]: Why did you change it in the first place?

File: tanner/utils/patterns.py (hunk 4 above)
Comment (rjt-gupta) [reply tanner/utils/patterns.py]: Template injection payloads can be matched with xss regex but vice versa is not possible.

File: tanner/emulators/template_injection.py (hunk 5 above)
Comment (rjt-gupta) [reply tanner/emulators/template_injection.py]: I tried but the implementation is old and I had to change a lot to make it work. This implementation is described in the latest docs.

https://docker-py.readthedocs.io/en/stable/containers.html#docker.models.containers.ContainerCollection.run

File: docker/docker-compose.yml (hunk 6 above)
Comment (rjt-gupta) [reply docker/docker-compose.yml]: Actually, my next task is replacing docker with aiodocker, so changing `docker_helper` is of no use as we will change it anyway. For now maybe lets just keep as is?

File: docker/tanner/template injection/Dockerfile (hunk 7 above)
Comment (rjt-gupta) [reply docker/tanner/template injection/Dockerfile]: It was to test if everything works perfectly or not.

File: tanner/emulators/template_injection.py (hunk 5 above)
Comment (rnehra01) [reply tanner/emulators/template_injection.py]: I think this is similar to cmd_exec emulator

File: tanner/utils/docker_helper.py (hunk 8 above)
Comment (rjt-gupta) [reply tanner/utils/docker_helper.py]: It is according to latest docker upgrade.

File: docker/docker-compose.yml (hunk 6 above)
Comment (rnehra01) [reply docker/docker-compose.yml]: I don't think it will be a lot of work, just see if a dockerfile is passed in method and build if it is else pull. And with aiodocker also, I'm sure similar methods will be present.

File: tanner/emulators/template_injection.py (hunk 5 above)
Comment (rjt-gupta) [reply tanner/emulators/template_injection.py]: I was difficult to execute complex payloads using old implementation.

File: docker/docker-compose.yml (hunk 6 above)
Comment (rjt-gupta) [reply docker/docker-compose.yml]: Cool, I'll give it a shot.

File: tanner/emulators/template_injection.py (hunk 5 above)
Comment (rnehra01) [reply tanner/emulators/template_injection.py]: Then maybe you can update the old implementation so that it can handle complex payloads.

Comment (rjt-gupta) [reply tanner/emulators/template_injection.py]: Okay! will look into it.

**Comment by rjt-gupta** [comment]
Ready.

**Comment by rnehra01** [comment]
It gives me `b'49'` for `{{7*7}}`. Can you remove the `b` and quotes?

File: tanner/emulators/template_injection.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        if patterns.TEMPLATE_INJECTION_TORNADO.match(payload):
+
+            tornado_template = 'import tornado\n' \
+                               'from tornado.template import Template\n' \
+                               'code = "%s"\n' \
+                               'result = tornado.template.Template(code)\n' \
+                               'template_injection_result = result.generate()\n' \
+                               'print(template_injection_result)' % payload
+
+            execute_result = self.docker_helper.docker_client.containers.run(
```
Comment (rnehra01) [inline tanner/emulators/template_injection.py]: What about this? What is the problem when you used docker_helper?

File: docker/tanner/template_injection/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+RUN apk -U --no-cache add \
+  python3
+
+RUN pip3 install --upgrade pip && \
+    pip3 install --no-cache-dir \
+    tornado \
+    jinja2 \
+    mako
+
+CMD ["python3", "-c", "import tornado"]
```
[the remark this answers is not in the document]
Comment (rjt-gupta) [reply docker/tanner/template_injection/Dockerfile]: Yes.

**Comment by rjt-gupta** [comment]
> It gives me `b'49'` for `{{7*7}}`. Can you remove the `b` and quotes?

I am decoding the output already and apparently `b'49'` as a whole is a string, dont know why it is getting encoded twice.

File: tanner/emulators/template_injection.py (hunk 9 above)
Comment (rjt-gupta) [reply tanner/emulators/template_injection.py]: I tried executing the vuln_code using `docker_helper.execute_cmd` method, basic payloads were working but strangely complex didnt worked. Thats why I used `docker_client` from `docker_helper` to execute commands which works perfectly.

Comment (rnehra01) [reply tanner/emulators/template_injection.py]: Can you specify the problem you're facing (using logs)?

**Comment by rjt-gupta** [comment]
Updated `docker_helper` to latest upgrade, updated lfi, cmd_exec, template_injection emulators also.
